### ISSUE TITLE
Missing types for `gapi.client.directory` 

### ISSUE THREAD

**Issue by MiroslavPetrik** [Maxim-Mazurok/google-api-typings-generator#652]
Hi

The [Admin SDK](https://developers.google.com/admin-sdk) contains multiple apis one of which is the [Directory API](https://developers.google.com/admin-sdk/directory/reference/rest).

There is [gapi.client.admin](https://www.npmjs.com/package/@maxim_mazurok/gapi.client.admin) which does not contain the Directory API calls. Note that it also includes `reports_v1` in the title.

There is no [gapi.client.directory](https://www.npmjs.com/package/@maxim_mazurok/gapi.client.directory) and there should be one.

I've checked the list of all service discovery docs and the `reports_v1` is marked as preferred. Also I've checked that the list contains the directory API service.
![image](https://user-images.githubusercontent.com/38855190/173813172-6afbd69a-efe8-4505-8003-d7f5c5e30ea4.png)

The `preferred` flag is used in practice https://github.com/Maxim-Mazurok/google-api-typings-generator/blob/master/src/app.ts#L1152 

So I think that not all the services are published on NPM and I would expect them to be there.

Edit:
I'm able to generate the `Directory API` types with command

```
npm start -- --url https://admin.googleapis.com/\\\$discovery/rest?version=directory_v1
```

The resulting folder is named `types/gapi.client.admin` which is in collision with the above mentioned npm package which has the `reports_v1` "version".

So I think the generator should be adjusted to generate APIs by id instead of name (see the screenshot above for difference between `id` and `name`).

Edit2 ((just to confirm)

In the generated readme it says:

```
## Installing

Install typings for Admin SDK API:

`npm install @types/gapi.client.admin@directory_v1 --save-dev`
```

But really it does not exist :

```
$ yarn add -D  @types/gapi.client.admin@directory_v1
➤ YN0000: ┌ Resolution step
➤ YN0016: │ @types/gapi.client.admin@npm:directory_v1: Registry failed to return tag "directory_v1"
➤ YN0000: └ Completed in 0s 974ms
➤ YN0000: Failed with errors in 0s 976ms
```

**Comment by Maxim-Mazurok** [comment]
Thank you for raising this issue!
You got it all quite right 👍 
Seems like we do have to generate and publish types for different ids as opposed to different names...
Shouldn't be too much of a trouble since everything except adding new types to DT is automated. Once we add new ids there - updates will be automatic.

And yes, seems like we need to change the readme template as well since we've changed versions to contain revisions.

I'll take a closer look at this, it was always not completely clear how google uses these ids and how consistent it is across different APIs. From your example, I understand that they published two completely different APIs under the same name but with different "version" parts of the id, which is not what I was expecting... Seems a bit like a misuse to me, most other APIs have `v3` or `v1beta1`... Anyway, it'd be nice to add support for deprecated or beta versions of APIs for a few who need them.

I also will have to check how long it'll take to generate, but judging by rough estimates, I think we still can get away with hourly updates.

Anyway, you did find a temporary workaround, just generate and add to your project for now, there should be some explanation here: https://github.com/Maxim-Mazurok/google-api-typings-generator/issues/216 and example here: https://github.com/Maxim-Mazurok/angular-google-calendar-typescript-example

And your PR would be very welcome as well! Essentially we just need to remove preferred/version filter, use id instead of name where needed (probably still use name in namespace for example. Actually, this makes me wonder how can you use admin directory and admin reports at the same time if they (should?) have the same namespace?). And remove `@`version from readme template.

Let me know if you have any questions in the meantime :)

**Comment by Maxim-Mazurok** [comment]
I see that it's `gapi.client.directory` https://developers.google.com/admin-sdk/directory/v1/quickstart/js and `gapi.client.reports` https://developers.google.com/admin-sdk/reports/v1/quickstart/js so they don't have namespace collision, which is good.

But the only promising way to get namespace is by looking at the canonicalName in JSON schema... And for `gapi.client.sheets` it's actually `Sheets`, so... we could just lowercase it, but seems sketchy. And Calendar API doesn't have canonicalName at all, but we could fallback to just name in this case which would produce the desired result.

If only we could de-minify gapi client that has all this logic - we could just use it instead of guessing... https://github.com/Maxim-Mazurok/gapi

**Comment by Maxim-Mazurok** [comment]
A quick update, I've started working on this, and added some thoughts in https://github.com/Maxim-Mazurok/google-api-typings-generator/blob/versions-support/TODO.md

TL;DR seems like we'll use ids from methods (like `admin.customer.devices.chromeos.commands.get`) to figure out what namespaces we have instead of relying on the API name. And Google API versions are crazy `v1p1beta1` i.e., so we'll have to publish them as separate packages.

**Comment by Maxim-Mazurok** [comment]
Another update: lots of stuff is being tested/refactored/done.
If you check out this commit https://github.com/Maxim-Mazurok/google-api-typings-generator/tree/1aabc4ed130a89e1e940c99a52624b4076073a23 you'll be able to generate directory API with correct namespaces, with `gapi.client.admin` stuff being separated from `gapi.client.directory` stuff, in the same typings file. This should better reflect the actual JS API.

**Comment by Maxim-Mazurok** [comment]
Quick update :) https://github.com/Maxim-Mazurok/google-api-typings-generator/tree/versions-support branch is now fully functional, I've already opened PR to add a new package to DT-tools, once I figure out how to mark old ones as deprecated - I'll open PR to DT as well, for now, please feel free to use that branch to generate correct types and tests for your APIs, testing and feedback will be highly appreciated!

**Comment by Maxim-Mazurok** [comment]
Hello @MiroslavPetrik! It's been a while, and I think that https://github.com/Maxim-Mazurok/google-api-typings-generator/pull/655 is ready now, please take a look and LMK if you can spot any issues.

I'll be merging it soon and opening PR to DT. I'll be keeping old packages for now with some deprecation messages in them. And adding new versioned packages as well. I ran tests and on the surface, everything seems to work. If you could try the new types generated by this PR on your actual project that'd be helpful, thank you!

**Comment by MiroslavPetrik** [comment]
Hey @Maxim-Mazurok 
I've generated the types from the PR and it now has the directory version:
![image](https://user-images.githubusercontent.com/38855190/183364984-c1d9b0ab-d7e1-4377-9262-25017857dbb8.png)

Also I've used the types in my project and it still builds successfully :)

Huge thank you for fixing this!

**Comment by Maxim-Mazurok** [comment]
My pleasure, I'll be merging it and opening PRs to DT, will let you know when it's published, thanks for bringing up the issue and testing!

**Comment by Maxim-Mazurok** [comment]
[@maxim_mazurok/gapi.client.admin-directory_v1](https://www.npmjs.com/package/@maxim_mazurok/gapi.client.admin-directory_v1) and [@maxim_mazurok/gapi.client.admin-reports_v1](https://www.npmjs.com/package/@maxim_mazurok/gapi.client.admin-reports_v1) are published now and will be auto-updated hourly, you can use them before https://github.com/DefinitelyTyped/DefinitelyTyped/pull/61722 for `@types` is merged
